apps/system/js/nfc_manager.js

```diff
@@ -200,134 +186,67 @@
      * @param {Object} msg - tech lost message
      */
     _handleTechLost: function nm_handleTechLost(msg) {
-      this._debug('Technology Lost: ' + JSON.stringify(msg));
+      this.debug('Technology Lost: ' + JSON.stringify(msg));
 
       window.navigator.vibrate([125, 50, 25]);
       window.dispatchEvent(new CustomEvent('nfc-tech-lost'));
 
       // Clean up P2P UI events
-      window.removeEventListener('shrinking-sent', this);
-      window.dispatchEvent(new CustomEvent('shrinking-stop'));
+      this._cleanP2PUI();
     },
 
-    /**
-     * Default event handler. Always listens for lockscreen-appopened,
-     * lockscreen-appclosed, screenchange. During P2P sharing flow it
-     * listens for shrinking-sent event dispatched from ShrinkingUI
-     * @memberof NfcManager.prototype
-     * @param {Event} event
-     */
-    handleEvent: function nm_handleEvent(evt) {
-      var state;
-      switch (evt.type) {
-        case 'lockscreen-appopened': // Fall through
-        case 'lockscreen-appclosed':
-        case 'screenchange':
-          if (!this.isActive()) {
-            return;
-          }
-          state = (ScreenManager.screenEnabled && !Service.locked) ?
-                    this.NFC_HW_STATE.ENABLE_DISCOVERY :
-                    this.NFC_HW_STATE.DISABLE_DISCOVERY;
-          if (state === this._hwState) {
-            return;
-          }
-          this._changeHardwareState(state);
-          break;
-        case 'shrinking-sent':
-          window.removeEventListener('shrinking-sent', this);
-          // Notify lower layers that User has acknowledged to send NDEF msg
-          this._dispatchP2PUserResponse();
-
-          // Stop the P2P UI
-          window.dispatchEvent(new CustomEvent('shrinking-stop'));
-          break;
+    _doNfcStateTransition: function(evt) {
+      var evtIdx = NfcHwEvents.indexOf(evt);
+      var state = NfcHwStateTable[this._hwState][evtIdx];
+      if (!state) {
+        this.debug('no transition from ' + this._hwState + '[' + evt + ']');
+        return;
       }
-    },
 
-    /**
-     * Basing on the new value of NFC Setting computes new NFC HW state
-     * and uses {@link NfcManager#_changeHardwareState} to set it
-     * @memberof NfcManager.prototype
-     * @param {boolean} enabled - NFC setting value
-     */
-    _nfcSettingsChanged: function nm_nfcSettingsChanged(enabled) {
-      this._debug('_nfcSettingsChanged, nfc.enabled: ' + enabled);
+      this._hwState = state;
+      this._notifyNfcStateChanged();
+      this._changeNfcHwState();
+    },
 
-      if (this.isActive() === enabled || this.isInTransition()) {
-        this._debug('_nfcSettingsChanged ignoring settings change');
-        return;
+    _notifyNfcStateChanged: function() {
```

Comment:
Yes, I agree that that these functions can be merged. I actually separated them because of the different actions they perform. ```this._notifyNfcStateChanged``` notifies external components (icon, nfc toggle via ```nfc.status``` setting) about the Nfc state change. ```this._changeNfcHwState``` asks Gecko to perform HW state change.